### ISSUE TITLE
0.6.5-1 fsync performance regression

### ISSUE THREAD

**Issue by hhhappe** [zfsonlinux/zfs#3780]
After upgrading to 0.6.5-1 fsync's are really slow. Ex.:

dd if=/dev/zero of=/zfs/t0 bs=1M count=128 conv=fsync
128+0 records in
128+0 records out
134217728 bytes (134 MB) copied, 11.0608 s, 12.1 MB/s

With 0.6.4-2 this was ~800MB/s. This is an 18 disk raidz2.

I've tried to create smaller raidz2, raidz, mirror and no redundancy. Only single disk and mirrored filesystems are giving more than 100MB/s.

Config:

CentOS 6.7
kernel: 2.6.32-573.3.1.el6.x86_64
1x E5-2640v3
128GB memory

**Comment by Bronek** [comment]
Which kernel version?

**Comment by hhhappe** [comment]
See last part of message.

**Comment by ryao** [comment]
This is likely a regression caused by the following:

https://github.com/zfsonlinux/zfs/commit/b39c22b73c0e8016381057c2240570f7af992def

That improves latencies, but likely had the effect of hurting though-put. We likely need to revisit this.

That said, I am surprised to hear that fsync was so fast with 0.6.4 in this setup. Unless sync=disabled is set, I would definitely not expect to see 800MB/sec. That level of performance should really only be possible with a SLOG device or flash storage because 800MB/sec of 1MB synchronous writes implies 800 IOPS. If you have a single 18-disk raidz2 vdev with mechanical disks, you should not be able to get more than 200 IOPS even with the best mechanical disks. 800 IOPS far exceeds the limit of what the disks typically are able to do, although it maybe is not impossible if we are laying things out in sequence on the disks.

What kinds of disks are these and what kind of controller(s) are in use?

**Comment by ryao** [comment]
We probably need to examine things, but just as a point of reference, here are figures from my workstation, which has a mirrored pair of Intel 730 SSDs, uses Linux 4.1 and everything is on ZFS with compression=lz4 and dedup=on:

```
desktop ~ # dd if=/dev/zero of=t0 bs=1M count=128 conv=fsync
128+0 records in
128+0 records out
134217728 bytes (134 MB) copied, 0.0401037 s, 3.3 GB/s
desktop ~ # rm t0 
desktop ~ # dd if=/dev/urandom of=t0 bs=1M count=128 conv=fsync
128+0 records in
128+0 records out
134217728 bytes (134 MB) copied, 8.82549 s, 15.2 MB/s
```

There is definitely room for improvement (15.2MB/sec is too slow), but the compression's zero detection meant that no data blocks actually had to be written.

**Comment by ryao** [comment]
I just did a flame graph analysis of the urandom case. It is CPU bound by the random number generator. If I adjust for that by copying the random output to a tmpfs, my system becomes bandwidth limited and I see realistic performance numbers:

```
desktop ~ # dd if=/tmp/t0 of=t0 bs=1M count=128 conv=fsync
128+0 records in
128+0 records out
134217728 bytes (134 MB) copied, 0.510636 s, 263 MB/s
```

**Comment by hhhappe** [comment]
@ryao: You just had me question my understanding of conv=fsync, so i did a quick strace. It only does one fsync as I expected.

This is plain LSI SAS hardware (Dell MD3060e and 9207-e8) and Seagate ST6000NM0034 disks. 

To get to 800MB/s I need to adjust zfs_vdev_sync_write_max_active up. Without fsync these 18 disks can at times reach 3GB/s.

**Comment by ryao** [comment]
@hhhappe I had not realized that `conv=fsync` only did one `fsync()` call. Thanks for the correction.

To what value are you setting `zfs_vdev_sync_write_max_active`?

**Comment by hhhappe** [comment]
Hmm, the zfs_vdev_sync_write_max_active increase helped for 0.6.3, but 0.6.4.2 performs without. It helped for files writes ending with fsync.

**Comment by behlendorf** [comment]
@hhhappe did the strace show all the time being spent in the call to fsync?

**Comment by hhhappe** [comment]
The strace spends almost all of the time in fsync. One 128x1M run gave ~200us per write and 12s in fsync.

**Comment by alexanderhaensch** [comment]
I can confirm this on my system too:

### with lz4 compression
#### with fsync
134217728 bytes (134 MB) copied, 0.175775 s, **764 MB/s**
#### without fsync
134217728 bytes (134 MB) copied, 0.072671 s, **1.8 GB/s**

### without compression
#### with fsync
134217728 bytes (134 MB) copied, 4.12551 s, **32.5 MB/s**
#### without fsync
134217728 bytes (134 MB) copied, 0.0699792 s, **1.9 GB/s**

### With a larger payload and without compression:
#### with fsync
10485760000 bytes (10 GB) copied, 27.7684 s, **378 MB/s**
#### without fsync
10485760000 bytes (10 GB) copied, 7.2256 s, **1.5 GB/s**

The large file is faster than the small one!

**Comment by behlendorf** [comment]
@hhhappe unfortunately I haven't been able to reproduce this problem.  fsync(2) performance appears unchanged in my testing between 0.6.4.2 and 0.6.5.1 when using a single disk pool under Centos 6.7.  It's certainly possible this problem only manifests itself for larger configurations be I haven't had a chance to test those yet.  Where is the effect more pronounced in your testing, large raidz2 configurations?

@alexanderhaensch do you have any data comparing performance with 0.6.4.2 vs 0.6.5.1.  We'd expect the fsync tests to be slower than the non-fsync ones, the question is how do the test results compare across releases.

**Comment by chjohnst** [comment]
I also ran tests to see if there is a diff, 0.6.5 actually performed
slightly faster (could be noise).  Are you certain this isn't a hardware
related issue?

**Comment by alexanderhaensch** [comment]
I ran some tests on different systems.
I found one system out of 5 that seems not affected, but all other do show the problem.
The Hardware is very different between the systems, so i do not see a red line here.

I will only show the conv=fsync results on uncompressed filesytems because they are the most significant.

First system LSI SAS2308
two disk, no redundancy
was:
uname -a
Linux iris 3.13.0-52-generic #86-Ubuntu SMP 
ZFS: Loaded module v0.6.4.2-1~trusty, ZFS pool version 5000, ZFS filesystem version 5
**191 MB/s**
then:
ZFS: Loaded module v0.6.5.1-1~trusty, ZFS pool version 5000, ZFS filesystem version 5
**37,7 MB/s**
-> thats a difference.

AMD Phenom X2 with onboard storage controller AMD SB8x0
raidz1 3 Disks

ZFS: Loaded module v0.6.4.2-r0-gentoo, ZFS pool version 5000, ZFS filesystem version 5
**76,8 MB/s**

ZFS: Loaded module v0.6.5-r1-gentoo, ZFS pool version 5000, ZFS filesystem version 5
**70,5 MB/s**
-> no differene here.

Then i have a HP Proliant Microserver.
Thats also an AMD system with a very similar Controller and Disks of the same type as in the system described before. It has 4 Disks and a smaller CPU.
Unfortunatly i have only the result for 0.6.5.1 because the system has autoupdates and i do not know how to downgrade zfs on ubuntu.
The transfer value is **4,2 MB/s**!
-> Thats too low i guess.

**Comment by hhhappe** [comment]
Here are some 0.6.5.1 numbers for striped and raidz2 for the regular disk counts (x: ). I noticed a difference in the first run right after create, so I included these as the first results for each disk count. For the very common 10 and 18 disk raidz2 the first run is faster, but not fast enough.

I will go and downgrade now and get the number for 0.6.4.2.

(stripe)

1: 134217728 bytes (134 MB) copied, 10.1796 s, 13.2 MB/s
1: 134217728 bytes (134 MB) copied, 0.786758 s, 171 MB/s

2: 134217728 bytes (134 MB) copied, 5.54385 s, 24.2 MB/s
2: 134217728 bytes (134 MB) copied, 0.999162 s, 134 MB/s

4: 134217728 bytes (134 MB) copied, 1.8845 s, 71.2 MB/s
4: 134217728 bytes (134 MB) copied, 1.54637 s, 86.8 MB/s

8: 134217728 bytes (134 MB) copied, 1.26512 s, 106 MB/s
8: 134217728 bytes (134 MB) copied, 1.28457 s, 104 MB/s

16: 134217728 bytes (134 MB) copied, 1.24824 s, 108 MB/s
16: 134217728 bytes (134 MB) copied, 0.901596 s, 149 MB/s


raidz2

3: 134217728 bytes (134 MB) copied, 5.89388 s, 22.8 MB/s
3: 134217728 bytes (134 MB) copied, 6.27775 s, 21.4 MB/s

4: 134217728 bytes (134 MB) copied, 5.22175 s, 25.7 MB/s
4: 134217728 bytes (134 MB) copied, 4.76767 s, 28.2 MB/s

6: 134217728 bytes (134 MB) copied, 4.02013 s, 33.4 MB/s
6: 134217728 bytes (134 MB) copied, 3.85458 s, 34.8 MB/s

10: 134217728 bytes (134 MB) copied, 2.59336 s, 51.8 MB/s
10: 134217728 bytes (134 MB) copied, 7.77536 s, 17.3 MB/s

18: 134217728 bytes (134 MB) copied, 2.89049 s, 46.4 MB/s
18: 134217728 bytes (134 MB) copied, 10.7906 s, 12.4 MB/s

**Comment by behlendorf** [comment]
@alexanderhaensch the most likely cause of the performance regression is commit b39c22b73c0e8016381057c2240570f7af992def.  Would it be possible for you to revert this change from 0.6.5.1 and see if performance improves?  Also are you using an external SSD based log device?

Here's what I think is happening.  Commit b39c22b73c0e8016381057c2240570f7af992def was designed to ensure that synchronous I/Os, like those written to the ZIL, get immediately sent to disk.  They're not allowed to queued in memory which would increase their latency.  This is great if you assume your log is on an SSD with low latency.  However, if your log is stored on the primary pool this is going to be terrible.  In that case you want to aggregate as much as you can since the write is going to be expensive!  We're going to need to distinguish these cases.

**Comment by alexanderhaensch** [comment]
I have a system without ZIL where i can test this. My patchset on 0.6.5.1 increases.. :) 
The system with ZIL is on 0.6.4.2 because it needs to be productive.

**Comment by alexanderhaensch** [comment]
In my case b39c22b73c0e8016381057c2240570f7af992def is not the culprit. The system withour b39c22b73c0e8016381057c2240570f7af992def performs exactly the same.

**Comment by behlendorf** [comment]
@alexanderhaensch in that case we may need to git bisect this to determine where the regression was introduced.

**Comment by alexanderhaensch** [comment]
I have to take my statement back. The system is i tested is slow on 0.6.4.2 and on 0.6.5.1..

**Comment by hhhappe** [comment]
Joggling to many things at the moment I don't have a complete set of results for the 0.6.4.2. Also, I noticed that the disk's write cache was not on. Apparently, that makes a huge difference on this test system (single MD3060 enclosure). I need to dig deeper here.

We have another system with 3x MD3060e enclosure connected to two servers. This is close to production mode, so I will not mess with pool setup (18 disk riadz2), but I will upgrade one server and see what happens.

Anyway, here are the 0.6.4.2 results for 18 disk raidz2 with 128MB and 1GB file size (disk write cache off).

Single MD3060e (same as before):

134217728 bytes (134 MB) copied, 1.23128 s, 109 MB/s
1073741824 bytes (1.1 GB) copied, 8.57962 s, 125 MB/s

S

**Comment by hhhappe** [comment]
Joggling to many things at the moment I don't have a complete set of results for the 0.6.4.2. Also, I noticed that the disk's write cache was not on. Apparently, that makes a huge difference on this test system (single MD3060 enclosure). I need to dig deeper here.

We have another system with 3x MD3060e enclosure connected to two servers. This is close to production mode, so I will not mess with pool setup (18 disk riadz2), but I will upgrade one server and see what happens.

Anyway, here are the 0.6.4.2 results for 18 disk raidz2 with 128MB and 1GB file size (disk write cache off).

Single MD3060e (same as before):

134217728 bytes (134 MB) copied, 1.23128 s, 109 MB/s
1073741824 bytes (1.1 GB) copied, 8.57962 s, 125 MB/s

So 10 times better than 0.6.5.1 and increasing file size does not help much.

3x MD3060e:

134217728 bytes (134 MB) copied, 0.227799 s, 589 MB/s
1073741824 bytes (1.1 GB) copied, 1.22117 s, 879 MB/s

A lot better.

**Comment by behlendorf** [comment]
@hhhappe thanks for getting this data.  Have you done the experiment of reverting b39c22b?  Did it help?

**Comment by hhhappe** [comment]
I haven't tried reverting b39c22b yet. Just tried 0.6.5.1 on the 3xMD3060e setup and it is even worse for the  raidz2 18x disk pool:

134217728 bytes (134 MB) copied, 26.0719 s, 5.1 MB/s
1073741824 bytes (1.1 GB) copied, 194.238 s, 5.5 MB/s

**Comment by hhhappe** [comment]
Okay reverting on the single MD3060e system did the trick. In fact, it is a lot faster than 0.6.4.2:

134217728 bytes (134 MB) copied, 0.205837 s, 652 MB/s
1073741824 bytes (1.1 GB) copied, 0.757666 s, 1.4 GB/s

Now, would that be a fix?

**Comment by alexanderhaensch** [comment]
@hhhappe is your latest report on the same configuration as two days ago? Is sync=disabled set? Back then you reported  between 12.4 MB/s and 46.4 MB/s..

**Comment by alexanderhaensch** [comment]
@behlendorf  Success! your commit 393ee23 solves the problem for me! 
134217728 bytes (134 MB) copied, 0.533725 s, 251 MB/s
Thats now on average 10times faster than on 0.6.5.1 and 0.6.4.2. I am not sure whats wrong with 0.6.4.2.
Unfortunatly i can only confirm this on gentoo, as i am still unable to patch on ubuntu.

**Comment by hhhappe** [comment]
@alexanderhaensch sync were the default.

**Comment by behlendorf** [comment]
@hhhappe @alexanderhaensch great news.  If reverting b39c22b fixed the issue then the patch in #3833 will address it as well.  It'll be in the next point release which shouldn't tag to long to finalize.

**Comment by behlendorf** [comment]
Resolved by 5592404 which will be cherry-picked in to 0.6.5.2 release.

**Comment by kernelOfTruth** [comment]
@behlendorf I've been wondering in the last weeks why updating the database of the recoll full desktop text search (http://www.lesbonscomptes.com/recoll/) was taking literally hours and giving me a hard time to complete during "closing" operation (fsync, merging, sorting, etc.) - thought it was "normal" to have this significantly worse performance compared to btrfs or ext4,

now it completes within a few minutes - it appears it was a significant regression

Thanks for the fix :+1:
